### ISSUE TITLE
Provide Escapes in code-block

### ISSUE THREAD

**Issue by skaller** [sphinx-doc/sphinx#4195]
I have a requirement to format code blocks for language Felix 
which uses TeX math embedded in code, particularly some important
operators are AMS Math symbols eg \circ for function composition.

With LaTeX  and minted package which uses Pygments I am able to 
write code like this:

```
\begin{minted}[bgcolor=felixbg,escapeinside=||]{felix}
pf |$\odot$| pg |$\odot$| ph
\end{minted}

which is spelled:

\begin{minted}[bgcolor=felixbg]{felix}
pf \odot pg \odot ph
\end{minted}
```

where  the `escapeinside` feature allows math directly in the code.

ReadTheDocs with Sphinx and Pygments works very nicely with MathJax
to provide the required formatting in ordinary text, but I need it in in
code-blocks and so I am requesting an enhancement which allows the
equivalent of minted's escapeinside feature. 

Substitutions would work at a pinch but are not ideal because the
math should really be written inline.

**Comment by jfbu** [comment]
There is no possibility of substitutions from inside code-blocks.

Can you clarify if your target is html or latex or ... ? The examples [there](http://felix-lang.org/share/src/web/advocate/mathinfelix.fdoc) are already converted for per symbol MathJax display and one does not get the Felix source code. My various tries to produce html from  a rst code-block containing `\alpha` failed with ``WARNING: Could not lex literal_block as "felix". Highlighting skipped.`` I also tried Unicode input ``val α = 10; ``

Regarding LaTeX, Sphinx uses an environment which already has ``\`` escape built-in so if one manages to produce something like ``\ensuremath{\odot}`` it will work. However as far as I know, Pygments does not allow escaping from its syntax highlighting except via comments. (maybe I am wrong).

Can you provide an example illustrating 

> ReadTheDocs with Sphinx and Pygments works very nicely with MathJax
> to provide the required formatting in ordinary text, but I need it in in
> code-blocks 

because it seems to imply you get Sphinx to use Pygments outside code-blocks to start with?

**Comment by skaller** [comment]
The primary target to make ReadTheDocs work online, so HTML. For paper like output I can already use LaTeX which is generally more powerful, if harder to use, than Sphinx.

The online examples at http://felix-lang.org/share/src/web/advocate/mathinfelix.fdoc are converted for MathJax by my custom webserver. The output you see there is the graphical design target. I am not asking for the code to be automatically displayed this way, as it is by my webserver, although that would be very nice. Rather, the code I posted in the first post is what I use with pdflatex, which uses the minted package, which in turn used Pygments. Minted package  provides a feature I would accept, which does require hand translation of the Felix code: I have to manually replace `\alpha`, for example, with `|$\alpha$|` and use the escapeinside feature of minted. Although this is a pain, it results in the code rendering correctly: the pdf output shows an greek alpha symbol in the program.

The problem with Sphinx is that its code-blocks feature does not allow this. 

On RTD I put conf.py to load sphinx.ext.mathjax and then the role :math: or :math:: stuff shows mathematics correctly using mathjax. So I can load mathjax, and it works just fine for RTD output using the math role, I just can't get that to go into the code-block.

Minted figured out how to put an escape into the Pygments output, and what's more, to do it using the worst language ever invented: TeX! So it is surely possible to make it work with Python in rST or Sphinx. 

I do not know how these tools function. Someone on Stackoverflow had a go at doing it though. Mathjax will do the formatting automatically if it can recognise the leadin/leadout escape symbols so what minted is doing is a providing some way to say "do not format this text as code, just leave it alone". Exactly how that works universally with Pygments I have no idea, but the escapeinside feature is not Felix specific, it works will all minted output. So it can be done, somehow, at the interface to Pygments.

It is possible, minted if fooling Pygments by telling it the escapeinside characters are comment delimiters and to leave the code alone. I don't know.

**Comment by jfbu** [comment]
As far as I understand from [this Pygments resolved issue](https://bitbucket.org/birkenfeld/pygments-main/issues/493/optionally-enable-escaping-to-latex),  ``escapeinside`` is a specific Pygments add-on feature for the LaTeX formatter.

I think what you want is an extension of this for adding MathJax to HTML output. If I am right in this understanding, this is more a Pygments issue than a Sphinx one.

On the pure LaTeX side, the  `sphinxVerbatim` environment already supports escapes. Especially if that escape does not attempt to change catcodes, and I just checked that minted's one has this aspect. Taking an example of the minted manual

```
\begin{minted}[escapeinside=||]{py}
def f(x):
y = x|\colorbox{green}{**}|2
return y
\end{minted}
```

If I modify `{green}{**}` to `{green}{*<four spaces>*}` I see spacing in output proving that the spaces did not  behave light in regular LaTeX context. In brief, they remained TeX-active. This naturally means that only the simplest of LaTeX code can go there. Regarding math mode however this is less problematic because the spaces will be discarded at low level by TeX.

Anyway, in Sphinx, we would not need even the ``|`` escapes, directly `\colorbox{green}{**}` would work. But Sphinx is very different because the LaTeX code is already Pygmentized.

To stop the digression, I am spending time on discussing LaTeX, because of ``minted`` featring in the discussion. But you were right to surmise that ``minted`` uses some general Pygments mechanism.

Unfortunately that mechanism seems to be specific for LaTeX Pygments formatter, and your main interest is into HTML formatter. I will investigate a bit more on Pygments side to see if I got things wrong, but I am of the opinion there is first and foremost a Pygments issue here.

**Comment by skaller** [comment]
Thanks, yes I'm doing that. I suspect a Pygments based solution is possible, I have emailed the author Georg Brandl for advice. For my purposes, the Felix lexer can probably do the required translation and emit the required mathmode, if Pygments can somehow be told not to process that text but emit it raw. In that case MathJax will pick it up automatically.

Yes, indeed, you're right, the escapeinside feature appears to be part of Pygments, and is leveraged by minted package for LaTeX .. and isn't available for HTML.

Thanks.

**Comment by tk0miya** [comment]
It sounds good. If pygments supports `escapeinside` option, we can use it on highlighting a node from `parsed-literal` directive!

**Comment by skaller** [comment]
I have been doing some experiments on RTD and it looks like what happens is that
this:
```
 :math:`\alpha^2` 
```
is replaced by this:
```
<span class="math">\(\alpha^2\)</span>
```

This is from viewing the page source of an actual RTD online document.
It is a bit hard to tell, because MathJax postprocesses the page,
and what comes out after the postprocessing is thousands of characters
of indecipherable gibberish :-)

MathJax usually triggers on ```\(``` and ```\)``` pairs, however
it can be prevented from seeing things in certain contexts.
Just putting 
```
\(\alpha^2\)
```
in the text does not work (it renders as written) so it looks like sphinx
is configuring MathJax to only look in class "math" spans or divs.
I can't be sure, I haven't found the configuration.

The hope is that Pygments can be tricked into emitting this text
which would probably require a Pygments lexer that can output
Math tokens or something, and then a Pygments formatter for HTML
that can translate that token into the above text.

The configuration for MathJax is probably somewhere in the
sphinx.ext.mathjax extension.

Somewhat unfortunately for me getting this to work requires expertise
in Pygment, Restructured Text processor, Sphinx, and Read-The_docs.

**Comment by jfbu** [comment]
I don't understand some of your observations regarding `\alpha` and the `class` tags. You can perhaps experiment easier with Sphinx by installing it locally and playing with `make html`, `make latex` on simple project.

From this `index.rst` source (and `extensions = ['sphinx.ext.mathjax']` in `conf.py`)

```rst
:math:`\alpha^2+\beta^2=\gamma^2`
```

I get html  like

```html
<p><span class="math">\(\alpha^2+\beta^2=\gamma^2\)</span></p>
```

and a working MathJax rendering.  The `class = "math"` things are removable (Sphinx can also output math via images, and then the CSS styling can help positioning the images.)

The MathJax configuration is done via

```html
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
```

**Comment by skaller** [comment]
Yes, you are confirming my observation. However the problem is 
to get mathjax to render maths in *Pygments* output. That is,
in *program code* not ordinary text.  We are trying to document
program code here, right? :) And my programming language allows
TeX identifiers as symbols.

In my webserver I use this configuration:

```
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
        skipTags: ["script","noscript","style","textarea"]
    }
  });
</script> 
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.jsconfig=TeX-AMS-MML_HTMLorMML">
</script>
```
What this means is MathJaX will *ignore* ```\(``` and ```\)``` delimiters inside the
specified tags.

According to the docs (on RTD):
```
skipTags: ["script","noscript","style","textarea","pre","code"]
This array lists the names of the tags whose contents should
 not be processed by tex2jax (other than to look for ignore/process 
classes as listed below). You can add to (or remove from) this list t
o prevent MathJax from processing mathematics in specific contexts.
```
which I think means those are the defaults. So if, in fact, Pygments output uses 
any of those tags, just inserting the ```\(``` and ```\)``` into the Pygments
output will not work. I'd be surprised if it *doesn't* use the pre tag. I do in
my webserver, but I changed the skipTags default so it does scan pre tags.
I mean, otherwise, I couldn't emit maths inside preformatted code.

I'm trying to connect with Pygments projects but Atlassian is screwed and won't register me.

**Comment by jfbu** [comment]
Well, it is hard for me to understand which observation of yours exactly my last comment confirmed.

As you indicate when one adds your config, it does work, for example here is some html

```html
<h1>Welcome to FOO’s documentation!<a class="headerlink" href="#welcome-to-foo-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="highlight-felix"><div class="highlight"><pre><span></span><span class="kd">val</span> <span class="n">\(\alpha\)</span> <span class="o">=</span> <span class="mi">10</span><span class="p">;</span>
<span class="k">fun</span> <span class="nf">f</span><span class="p">(</span><span class="n">x</span><span class="p">:</span><span class="kt">int</span><span class="p">)</span><span class="o">=&gt;</span> <span class="n">x</span> <span class="o">*</span> <span class="n">x</span><span class="p">;</span>
```

which was initially obtained from Sphinx `make html` with this source

```rst
Welcome to FOO's documentation!
===============================

.. highlight:: felix

::

  val a = 10;
  fun f(x:int)=> x * x;

  println$ "Hello World";
```

and then I went to replace manually the ``a`` by ``\(\alpha\)`` inside html and I added your 

```html
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
      tex2jax: {
      skipTags: ["script","noscript","style","textarea"]
      }
      });
    </script> 
```

to the existing

```html
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
```

and MathJax works as expected.

The problem is that I put ``\alpha`` in index.rst, the Pygments Felix lexer refuses it anyhow. Same with U+03B1 ``α``. We are back to initial issue that Pygments Felix lexer does not know how to output MathJax syntax on specific tokens.

Regading ``parsed-literal`` directive in Sphinx, we can use it indeed as 

```rst
.. parsed-literal::

   val :math:`\alpha` = 10;
   fun f(x:int)=> x * x;

   println$ "Hello World";
```

and together with your

```html
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
      tex2jax: {
      skipTags: ["script","noscript","style","textarea"]
      }
      });
    </script> 
``` 

the MathJax works. Sadly though there is no Pygments highlighting here but this expected as only via ``code-blocks`` do we get Pygmentization at Sphinx.

Also Sphinx's ``parsed-literal`` has an annoying aspect that it sometimes falls back to ``code-block`` behaviour, for example if I replace ``:math:`\alpha` `` in the above by ``a``, this is the case. We can add a comment  ``// *comment*`` where the emphasis will force Sphinx to accept this falls in to ``parsed literal`` category.

Anyway, the problem here is how to get Pygments to accept MathJax escapes. The extra config for ``pre`` and ``code`` is easily added at Sphinx user side.

**Comment by jfbu** [comment]
> Anyway, the problem here is how to get Pygments to accept MathJax escapes. The extra config for pre and code is easily added at Sphinx user side.

or rather the Pygments Felix lexer should recognize tokens (like U+03B1  or `\alpha`) which in output formatters will need some kind of math rendering: MathJaX for html, (or images), and TeX math mode for LaTeX.

**Comment by skaller** [comment]
Right, I know the Felix lexer doesn't recognise TeX. Even if it did, I'm not sure if there's
a way to output the required TeX without *also* doing a new style file. Not sure.
In *my* local LaTeX documents, Pygments is using my upgraded lexer,
RTD is using a very out of date one. But then, even my upgraded lexer still
doesn't recognise TeX.

Remember I'm not at liberty to generate the docs myself, ReadTheDocs is doing it.
So when you added my MathJax.Hub.Config, you did it locally, but i do not
know how to do it so RTD accepts it.

There's also an issue doing with Pygments, first I have to get it right.
Then I have to get the patches incorporated into the official Pygments,
wait for it to be release, and then wait for RTD to upgrade their Pygments.
This could take months. In fact unless Atlassian fixes their system it 
might never happen because I cannot register with them, and Pygments
is on their BitBucket system.

So maybe there's a way to get RTD to use a custom lexer. Its not just a
sphinx issue, I can do more locally than I can on RTD. Exactly what I 
can and can't do I don't know.

When I load the LaTeX generated PDF from RTD I get good colours.
But the HTML version is missing most of the colours. Whether that's
Pygments issue or sphinx issue of a lexer issue or what I don't know :-)

**Comment by skaller** [comment]
Ok, so on a *local* version of sphinx with a modified Pygments lexer, I can get
TeX symbols rendered by MathJax. The lexer mod if incomplete w.r.t. Felix actual
rules. Add this to the start of the lexer:

```
    def get_tokens_unprocessed(self,text):
      for index,token,value in RegexLexer.get_tokens_unprocessed(self,text):
        if value[0]=="\\":
          value = "\\(" + value + "\\)"
        yield (index,token,value)
```

and make the identifiers entry  in the lexer read:
```
            # Identifiers
            (r"[a-zA-Z_][a-zA-Z0-9_'-]*|\\[A-Za-z]+", Name),
```

All that remains is to hack the HTML with this *before* the mathjax inclusion:
```
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
        skipTags: ["script","noscript","style","textarea"]
    }
  });
</script> 
```

I'm not sure how to get this to happen from conf.py yet. Unfortunately it seems
pip has installed a frozen sphinx and I have no access to the sphinx.ext.mathjax
file. I need that to modify it unless there is a better way to add the extra Javascript?

**Comment by skaller** [comment]
BTW: the above is a hack because it will screw up LaTeX output. The correct way to do this in Pygments is to add a token type TeXMath and then have the formatter for HTML and LaTeX do the translation.

This would impact Ocaml too. In Ocaml type variables are written 'a, 'b etc, but they should be typeset as as Greek letters alpha, beta, etc.

**Comment by jfbu** [comment]
Are you sure you need it before MathJax inclusion? Brief testing seems to work even if after. To obtain that (after) inclusion, put the following as file ``layout.html`` in ``_templates/`` directory.

```
{% extends "!layout.html" %}
{% block extrahead %}
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
           skipTags: ["script","noscript","style","textarea"]
      }
      });
    </script>
{% endblock %}
```

Please check [Sphinx documentation](http://www.sphinx-doc.org/en/stable/templating.html?highlight=extra%20css#jinja-sphinx-templating-primer) on templating.

I tested with ``html_theme = 'sphinx_rtd_theme'`` in ``conf.py`` (and ``templates_path = ['_templates']``, and the extra head script is included, and indeed MathJax works as expected even within ``<pre>..</pre>``.

**Comment by jfbu** [comment]
ah indeed it ends up _before_ when using `sphind_rtd_theme`; but with `alabaster` it will be included _after_. Nevertheless the MathJax does work there too in ``<pre>..</pre>``.

**Comment by skaller** [comment]
From this link:

http://docs.mathjax.org/en/latest/configuration.html#using-in-line-configuration-options

it explains the inline configuration options must be given first. Interesting you found this not to be the case... perhaps it depends on which options.


Using in-line configuration options
The second way to configure MathJax is through in-line configuration, which puts the configuration options within the web page itself. The use of in-line configuration with MathJax requires two separate <script> tags: one for specifying the configuration settings and one for loading of MathJax. Because MathJax starts its configuration process as soon as it is loaded, the configuration script must come before the script tag that loads MathJax.js itself. You do this by including a <script> with type="text/x-mathjax-config" whose content will be run when MathJax performs its configuration. Generally, this script will include a MathJax.Hub.Config() call to perform MathJax configuration, but it can also include other MathJax commands, such as registering signal actions, or any JavaScript commands that you want. You can have as many such script tags as you need, and MathJax will process them in the order in which they appear in the document.

**Comment by skaller** [comment]
BTW: I can say: with the modified Felix lexer for Pygments, Pygments screws up the LaTeX output if it uses "escapeinside". Without escapeinside, just plain \alpha in the code, it just prints the ```\(``` and ```\)``` delimiters. It also prints $.  It seems Pygments is incorrectly using $ delimiters for math (LaTeX recommends using ```\(``` and ```\)```). I think this means my "hack" to the lexer won't fly. It has to be done the right way, lexing to a Math token or something, and then have Pygments do the translation in the formatter.

**Comment by skaller** [comment]
Oops above I meant ```\(``` and ```\)``` delimiters.

**Comment by jfbu** [comment]
indeed in LaTeX code-blocks, as produced by Sphinx, using `\(` and `\)` is correct way to escape to math mode. I should have said that earlier (rather than `\ensuremath` thing). Notice that  contrarily to `minted` context mentioned at top of thread, produced LaTeX mark-up needs no extra escape. Simply `\(` and `\)`. In short, you need same mark-up as for HTML MathJax ready output. (well not quite, if you want coloring too on symbols)

**Comment by skaller** [comment]
Right, but this doesn't work using *minted* package inside LaTeX,
rather than using Sphinx to create a LaTeX document. I tried it.
It bugs out LaTeX.

Just to be clear, the pipelines are, roughly:
```
1. reST :  Pygments -> Sphinx -> HTML
2. reST :  Pygments -> Sphinx -> LaTeX -> PDF
3. LaTeX : Pygments -> Minted -> LaTeX -> PDF
```

The first pipeline is what RTD is using via Sphinx to make HTML from reST documents.
The second is what RTD is using via Sphinx to make LaTeX and then PDF from reST documents.
The third is what I use to produce PDF from LaTeX documents.

Roughly, full LaTeX is much better for professional quality off line docs
such as an book. However the resulting PDF, whilst is is click-navigable,
is not nearly as good for interactive browsing as RTD. So I don't have that
much interest in the second pipeline.  The other big advantage of sphinx
with RTD is that it is really easy to edit and deploy docs, without too much
arcane knowledge which is definitely required for full LaTeX use. And LaTeX
isn't that much chop with non-maths either. Just edit, commit, and the docs
appear by magic :)

The point is that changes to Pygments have to work on all three pipelines.

This isn't just a Felix issue either. Ocaml also requires TeX symbols
to be formatted correctly. And realistically a lot of other languages could
do with nicer symbols such as nice arrows instead of AsciiArt.

It does look like Sphinx is easily capable of handling the integration.
I am guessing mods to Pygments won't be too hard either.
The main hassle is going to be deployment of the mods.

**Comment by skaller** [comment]
Hmmm. I have made an extension _static/felixmathjax.py from sphinx/ext/mathjax.py to load _static/felixmathjax.js, which is the MaxJax.Hub.Config code in previous posts. The modified felixmathjax.py is correctly loaded and used from the conf.py file.

I also modified my lexer. I had to destructively update my system library for that.

This is generated as expected:
```
<script src="_static/felixmathjax.js"></script>
```
 in the HTML but does not work.

If I replace it with the actual javascript from that file  it works:
```
MathJax.Hub.Config({
    tex2jax: {
        skipTags: ["script","noscript","style","textarea"]
    }
  });
```
I have no idea what is going on. All I actually need to do is 
remove the default "pre" tag from the skipTags list. Unfortunately
I don't know how to emit HTML/JS into the generated file.
The app.add_javascript() function in application.py is badly named,
it adds a file, not actual script.

**Comment by jfbu** [comment]
Isn't the [template override method](https://github.com/sphinx-doc/sphinx/issues/4195#issuecomment-340872142) working for you? Are you doing other things in your python code apart from adding the Javascript to generated HTML?

**Comment by skaller** [comment]
ah sorry, i haven't tried that method. Will try now.

**Comment by skaller** [comment]
Yeah, that works despite being added in the wrong place. That worries me but it seems to work.

So the next problem is, how to get my new Felix lexer loaded by Pygments. The trick there was just to emit the token "Generic" in the lexer, and then, munge get_tokens_unprocessed() to add ```\(```, ```\)``` around the math text. May mess up the LaTeX output but that's another issue.

**Comment by skaller** [comment]
There is another problem, the colours on my local system are fine. But on RTD they're barely visible, if they exist at all, except for the really bad choice of red for all the wrong things. I tried different styles and none work. It seems like RTD is messing about with the opaqueness. Its not just Felix. C++ code is affected too. So is Python. The colours are there, they're just not strong enough.

**Comment by skaller** [comment]
I think this is more a Python question than anything, although possibly it should be an enhancement request.

I need to add a custom pygments lexer. There is an API call sphinx.add_lexer(lexer, alias) however the lexer must already be in sphinx.highlighting. So I need to extend sphinx.highlighting, which I propose to do by brute force: I will change the python path to load MY copy of highlighting from _static. It will be a copy of the one in the repository with one extra line which imports lexers from _static *as well*, I guess ```from  custom_lexers import *``` would work???

I guess a better solution in the long run would be ```sphinx.add_custom_lexer```.

Note I cannot modify sphinx or pygments *except* via the conf.py file.

**Comment by skaller** [comment]
Ok this seems to work:

```
sys.path.insert(0, os.path.abspath('./_static'))
import custom_lexers.xfelix

from sphinx.highlighting import lexers
lexers['xfelix']= custom_lexers.xfelix.XFelixLexer()
```

**Comment by skaller** [comment]
Final result, needs tweaking but the interfacing now works:
![mathinfelix](https://user-images.githubusercontent.com/431744/32813860-6ef3f6de-ca00-11e7-86c6-bec753956d30.jpeg)

Although not faithful to the real Felix encoding at the moment it is good enough
for docs. I used a custom lexer xfelix which shows TeX symbols, and will also do
an updated lexer for felix which shows what you see in your editor.

**Comment by skaller** [comment]
BTW: Doesn't work on RTD, the _templates/layout.html file isn't included. The lexer is running though.

**Comment by skaller** [comment]
http://felix-tutorial.readthedocs.io/en/latest/hello.html

**Comment by jfbu** [comment]
Good that you apparently cleared-up obstacles! I think people visiting the tutorial will really need to see and be able to copy/paste the felix source code prior to any MathJaX rendering. I understand this is intended?

**Comment by skaller** [comment]
At the moment, that is doubly impossible. First, the math codes there represent real Felix code BUT the encoding is a cheat. Even on flx_web, which translates *real* Felix code, the output TeX is not quite the same as the Felix code. Whether you can actually copy HTML at all is another question.

I am looking now at TWO lexers, one which shows the code "as you see it in your dumb editor" and one you would see if it were typeset. Just trying to figure out how to move the common stuff out of each document _static (there are currently FOUR separate documentation projects and will probably end up being more).

The trouble is there are some key operators that require TeX. Function composition is spelled ```\circ``` and reverse composition is spelled ```\odot```.  Ascii just doesn't have enough operators :-)

On flx_web I could turn off MathJaX. That would be too hard with Sphinx on RTD. I'd need to add special buttons or whatever. At home .. I just turn off my internet connection, then MathJax can't load :-)

**Comment by skaller** [comment]
Just FYI: TeX uses {} for grouping, and does not respect (). Felix of course uses () for grouping, and {} is actually used in expressions for lambdas. Felix turns "math mode" on with `\( \)`. I forget whether that actually prints. But when we see (), we have to add {} around it for TeX. Also \letters is a valid Felix identifier *outside* math mode, but we have to add `\(\)` around it so MathJax see it.

Also, even in math mode, multi-character identifiers must be treated atomically as identifiers, so only single characters are typeset slanted as math. 

In other words, the output text *before* translation to HTML is already not valid Felix.
TeX really wasn't designed to typeset programming languages. It just has a lot of extra operators which are seriously missing from ASCII. Unfortunately Unicode isn't available on many editors, isn't well handled by many tools (regexps etc), doesn't typeset correctly, and is also seriously lacking suitable operators. Even if you steal APL character set, its good for arrays and not much else.
